### PR TITLE
Support for dat3m

### PR DESCRIPTION
**Pull request by apaolillo**
D'Artagnan and the Three Musketeers (``dat3m''), besides being a chef-d'oeuvre authored by Alexandre Dumas, is also a Bounded Model Checker for concurrent C programs to be validated against Weak Memory Models hardware architectures.
Its source code can be found on GitHub:
https://github.com/hernanponcedeleon/Dat3M.git

Integrating dat3m into benchkit will provide several benefits, among others:
- Benchmark several options of the model checker, like the underlying SMT solvers, the number of unrolled loops, or why not a future support for multiple threads?
- Use dat3m into larger toolchains allowing also to evaluate the performance of verified programs on different kind of hardware.

Notice the dat3m support depends on the `pythainer` package, allowing to manipualte pythonically the build of Docker images and the run of associated containers, as we run the verification inside Docker containers. Obviously, Docker also needs to be installed.

### CODE REVIEW

File: examples/dat3m/campaign_dat3m.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    campaign = CampaignCartesianProduct(
+        name="dat3mlocks",
+        benchmark=dat3m_bench,
+        nb_runs=nb_runs,
+        variables={
+            "memory_model": ["aarch64"],
+            "target_arch": ["arm8"],
+            "lock_name": list(dat3m_bench.dat3m_locks),
+            "bound": [None, 1, 3],
```
Comment (hernanponcedeleon) [inline examples/dat3m/campaign_dat3m.py:48]: I assume `None` avoids explicitly adding the `--bound=...` option to dat3m cmd. In such a case, the default value is 1, and thus having `None` and `1` is redundant.

File: examples/dat3m/kit/bench.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            lock_name=lock_name,
+            bound=bound,
+            **kwargs,
+        )
+
+        # Generate dat3m command line using variables
+        command_head = [
+            "java",
+            "-jar",
+            "dartagnan/target/dartagnan.jar",
```
Comment (hernanponcedeleon) [inline examples/dat3m/kit/bench.py:154]: This is not just intended to be run using the docker containers, right? The user could run benchkit from host if dat3m is installed. If so, maybe you should prefix every file/folder related to dat3m with `DAT3M_HOME`

File: examples/dat3m/kit/dat3m.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    DockerfileDockerBuilder,
+    UbuntuDockerBuilder,
+)
+from pythainer.builders.utils import project_git_clone
+from pythainer.examples.builders import get_user_builder
+
+from benchkit.utils.git import clone_repo
+
+
+def vsyncer_builder() -> DockerBuilder:
```
Comment (hernanponcedeleon) [inline examples/dat3m/kit/dat3m.py:15]: Why is vsyncer part of dat3m.py?

File: examples/dat3m/campaign_dat3m.py (hunk 1 above)
Comment (apaolillo) [reply examples/dat3m/campaign_dat3m.py:48]: I see. I will keep it supported as it is but I can remove it from the campaign.

File: examples/dat3m/kit/bench.py (hunk 2 above)
Comment (apaolillo) [reply examples/dat3m/kit/bench.py:154]: indeed, but i need to support retrieving value in environment variable through the comm layer (either docker or the host). That can be done.

File: examples/dat3m/kit/dat3m.py (hunk 3 above)
Comment (apaolillo) [reply examples/dat3m/kit/dat3m.py:15]: I thought to support both ways to use dat3m. But maybe vsyncer deserves its own file/example directory.

**Review by hernanponcedeleon — APPROVED** [review]

**Comment by apaolillo** [comment]
thanks for the review!

**Pull request merged**